File: LR1/Array_Lists_Stack/DoublyLinkedList.cpp
#include "DoublyLinkedList.h"
#include <fstream>
#include <sstream>

// Создание нового узла
LList* createNodeD(string value) {
    LList* newNodeD = new LList;
    newNodeD->data = value;
    newNodeD->next = nullptr;
    newNodeD->pred = nullptr;
    return newNodeD;
}

// Добавление в начало (head)
void LaddFront(LList*& head, LList*& tail, string value) {
    LList* NodeD = createNodeD(value);
    if (!head) {
        head = tail = NodeD;
    }
    else {
        NodeD->next = head;
        head->pred = NodeD;
        head = NodeD;
    }
}

// Добавление в конец (tail)
void LaddBack(LList*& head, LList*& tail, string value) {
    LList* NodeD = createNodeD(value);
    if (!tail) {
        head = tail = NodeD;
    }
    else {
        tail->next = NodeD;
        NodeD->pred = tail;
        tail = NodeD;
    }
}

// Добавление после узла с указанным значением
void LaddAfter(LList*& head, LList*& tail, string afterValue, string newValue) {
    LList* current = head;
    while (current && current->data != afterValue)
        current = current->next;

    if (!current) {
        cout << "Элемент " << afterValue << " не найден.\n";
        return;
    }

    LList* NodeD = createNodeD(newValue);
    NodeD->pred = current;
    NodeD->next = current->next;

    if (current->next)
        current->next->pred = NodeD;
    else
        tail = NodeD; // вставка после хвоста

    current->next = NodeD;
}

// Добавление перед узлом с указанным значением
void LaddBefore(LList*& head, LList*& tail, string beforeValue, string newValue) {
    LList* current = head;
    while (current && current->data != beforeValue)
        current = current->next;

    if (!current) {
        cout << "Элемент " << beforeValue << " не найден.\n";
        return;
    }

    LList* NodeD = createNodeD(newValue);
    NodeD->next = current;
    NodeD->pred = current->pred;

    if (current->pred)
        current->pred->next = NodeD;
    else
        head = NodeD; // вставка перед головой

    current->pred = NodeD;
}

// Удаление из начала
void LreFront(LList*& head, LList*& tail) {
    if (head == nullptr) {
        cout << "Список пуст\n";
        return;
    }
    LList* temp = head;
    head = head->next;
    if (head != nullptr)
        head->pred = nullptr;
    else
        tail = nullptr; // список стал пуст
    delete temp;
}

// Удаление из конца
void LreBack(LList*& head, LList*& tail) {
    if (tail == nullptr) {
        cout << "Список пуст\n";
        return;
    }
    LList* temp = tail;
    tail = tail->pred;
    if (tail != nullptr)
        tail->next = nullptr;
    else
        head = nullptr; // список стал пуст
    delete temp;
}

// Удаление элемента после заданного значения
void LreAfter(LList*& head, LList*& tail, string afterValue) {
    LList* cur = head;
    while (cur && cur->data != afterValue)
        cur = cur->next;

    if (!cur || !cur->next) {
        cout << "Невозможно удалить элемент после " << afterValue << endl;
        return;
    }

    LList* toDel = cur->next;
    cur->next = toDel->next;

    if (toDel->next)
        toDel->next->pred = cur;
    else
        tail = cur;

    delete toDel;
}

// Удаление элемента перед заданным значением
void LreBefore(LList*& head, LList*& tail, string beforeValue) {
    LList* cur = head;
    while (cur && cur->data != beforeValue)
        cur = cur->next;

    if (!cur || !cur->pred) {
        cout << "Невозможно удалить элемент перед " << beforeValue << endl;
        return;
    }

    LList* toDel = cur->pred;
    if (toDel->pred)
        toDel->pred->next = cur;
    else
        head = cur;

    cur->pred = toDel->pred;
    delete toDel;
}

// Удаление по значению
void LreValue(LList*& nd, LList*& tail, string value) {
    if (nd == nullptr) {
        cout << "Список пуст\n";
        return;
    }

    LList* current = nd;
    while (current != nullptr) {
        if (current->data == value) {
            if (current == nd)
                LreFront(nd, tail);
            else if (current == tail)
                LreBack(nd, tail);
            else {
                current->pred->next = current->next;
                current->next->pred = current->pred;
                delete current;
            }
            cout << "Удален элемент со значением " << value << endl;
            return;
        }
        current = current->next;
    }
    cout << "Элемент " << value << " не найден\n";
}

// Поиск по значению — возвращает индекс (или -1, если не найден)
int Lsearch(LList* nd, string value) {
    int index = 0;
    LList* current = nd;
    while (current != nullptr) {
        if (current->data == value)
            return index;
        current = current->next;
        index++;
    }
    return -1; // не найден
}

// Вывод списка от начала к концу
void printForward(LList* nd) {
    cout << "Список (прямой): ";
    LList* current = nd;
    while (current != nullptr) {
        cout << current->data << " ";
        current = current->next;
    }
    cout << endl;
}

// Вывод списка от конца к началу
void printBackward(LList* tail) {
    cout << "Список (обратный): ";
    LList* current = tail;
    while (current != nullptr) {
        cout << current->data << " ";
        current = current->pred;
    }
    cout << endl;
}

// Очистка памяти
void Lclear(LList*& nd) {
    LList* current = nd;
    while (current != nullptr) {
        LList* next = current->next;
        delete current;
        current = next;
    }
    nd = nullptr;
}

File: LR1/NaGO/Array_Lists_Stack/structures/doublylist/doubly_linked_list.go
package doublylist

import "fmt"

// LList — узел двусвязного списка
type LList struct {
	Data string
	Next *LList
	Pred *LList
}

// Создание нового узла
func createNodeD(value string) *LList {
	return &LList{
		Data: value,
		Next: nil,
		Pred: nil,
	}
}

// Добавление в начало (head)
func LaddFront(head **LList, tail **LList, value string) {
	node := createNodeD(value)
	if *head == nil {
		*head = node
		*tail = node
	} else {
		node.Next = *head
		(*head).Pred = node
		*head = node
	}
}

// Добавление в конец (tail)
func LaddBack(head **LList, tail **LList, value string) {
	node := createNodeD(value)
	if *tail == nil {
		*head = node
		*tail = node
	} else {
		(*tail).Next = node
		node.Pred = *tail
		*tail = node
	}
}

// Добавление после узла с указанным значением
func LaddAfter(head **LList, tail **LList, afterValue, newValue string) {
	current := *head
	for current != nil && current.Data != afterValue {
		current = current.Next
	}
	if current == nil {
		fmt.Println("Элемент", afterValue, "не найден.")
		return
	}

	node := createNodeD(newValue)
	node.Pred = current
	node.Next = current.Next

	if current.Next != nil {
		current.Next.Pred = node
	} else {
		*tail = node // вставка после хвоста
	}
	current.Next = node
}

// Добавление перед узлом с указанным значением
func LaddBefore(head **LList, tail **LList, beforeValue, newValue string) {
	current := *head
	for current != nil && current.Data != beforeValue {
		current = current.Next
	}
	if current == nil {
		fmt.Println("Элемент", beforeValue, "не найден.")
		return
	}

	node := createNodeD(newValue)
	node.Next = current
	node.Pred = current.Pred

	if current.Pred != nil {
		current.Pred.Next = node
	} else {
		*head = node // вставка перед головой
	}
	current.Pred = node
}

// Удаление из начала
func LreFront(head **LList, tail **LList) {
	if *head == nil {
		fmt.Println("Список пуст")
		return
	}
	temp := *head
	*head = (*head).Next
	if *head != nil {
		(*head).Pred = nil
	} else {
		*tail = nil
	}
	_ = temp // GC сам освободит
}

// Удаление из конца
func LreBack(head **LList, tail **LList) {
	if *tail == nil {
		fmt.Println("Список пуст")
		return
	}
	temp := *tail
	*tail = (*tail).Pred
	if *tail != nil {
		(*tail).Next = nil
	} else {
		*head = nil
	}
	_ = temp
}

// Удаление элемента после заданного значения
func LreAfter(head **LList, tail **LList, afterValue string) {
	cur := *head
	for cur != nil && cur.Data != afterValue {
		cur = cur.Next
	}
	if cur == nil || cur.Next == nil {
		fmt.Println("Невозможно удалить элемент после", afterValue)
		return
	}
	toDel := cur.Next
	cur.Next = toDel.Next
	if toDel.Next != nil {
		toDel.Next.Pred = cur
	} else {
		*tail = cur
	}
	_ = toDel
}

// Удаление элемента перед заданным значением
func LreBefore(head **LList, tail **LList, beforeValue string) {
	cur := *head
	for cur != nil && cur.Data != beforeValue {
		cur = cur.Next
	}
	if cur == nil || cur.Pred == nil {
		fmt.Println("Невозможно удалить элемент перед", beforeValue)
		return
	}
	toDel := cur.Pred
	if toDel.Pred != nil {
		toDel.Pred.Next = cur
	} else {
		*head = cur
	}
	cur.Pred = toDel.Pred
	_ = toDel
}

// Удаление по значению
func LreValue(head **LList, tail **LList, value string) {
	if *head == nil {
		fmt.Println("Список пуст")
		return
	}
	current := *head
	for current != nil {
		if current.Data == value {
			if current == *head {
				LreFront(head, tail)
			} else if current == *tail {
				LreBack(head, tail)
			} else {
				current.Pred.Next = current.Next
				current.Next.Pred = current.Pred
			}
			fmt.Println("Удален элемент со значением", value)
			return
		}
		current = current.Next
	}
	fmt.Println("Элемент", value, "не найден")
}

// Поиск по значению — возвращает индекс
func Lsearch(head *LList, value string) int {
	index := 0
	current := head
	for current != nil {
		if current.Data == value {
			return index
		}
		current = current.Next
		index++
	}
	return -1
}

// Вывод списка от начала к концу
func PrintForward(head *LList) {
	fmt.Print("Список (прямой): ")
	current := head
	for current != nil {
		fmt.Print(current.Data, " ")
		current = current.Next
	}
	fmt.Println()
}

// Вывод списка от конца к началу
func PrintBackward(tail *LList) {
	fmt.Print("Список (обратный): ")
	current := tail
	for current != nil {
		fmt.Print(current.Data, " ")
		current = current.Pred
	}
	fmt.Println()
}

// Очистка списка
func Lclear(head **LList) {
	for *head != nil {
		*head = (*head).Next
	}
}
